### PR TITLE
Improved behavior of the installer around the handling of Netdata Cloud support.

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

This adds two specific improvements to the installer related to the Netdata Cloud support:

* It adds a command-line option `--require-cloud` which will explictly request that the build fails if Netdata Cloud support can't be built. This is mostly intended for use in our CI to test the build functionality around Netdata Cloud, but it is also potentially of interest to some users, so ti's properly documented. It is mutually exclusive with `--disable-cloud`, if both are specified, only the first one is honored (and an error message is logged).
* It adds support code for deferring non-fatal error messages so that they get printed at the end of the install process (if the install otherwise completes successfully), and adds this functionality to the handling code for libmosquitto, LWS, and the go.d plugin (so that you end up with a nice summary of errors at the end of the install process indicating a lack of support for Netdata Cloud or for the go.d plugin.

##### Component Name

area/packaging

##### Description of testing that the developer performed

Minimal verification so far, will mark ready for review once Ive finished thorough testing.

### CODE REVIEW

File: netdata-installer.sh [hunk 1]
```diff
@@ -40,6 +40,24 @@ fi
...
+defer_error() {
+  NETDATA_DEFERRED_ERRORS="${NETDATA_DEFERRED_ERRORS}\n* ${1}"
+}
+
+print_deferred_errors() {
+  if [ -n "${NETDATA_DEFERRED_ERRORS}" ] ; then
+    echo >&2
+    echo >&2 "The following non-fatal errors were encountered during the installation process:"
+    # shellcheck disable=SC2059
+    printf >&2 "${NETDATA_DEFERRED_ERRORS}"
```
Comment (prologic) [inline netdata-installer.sh]:
```suggestion
    printf >&2 "%s" "${NETDATA_DEFERRED_ERRORS}"
```

File: netdata-installer.sh [hunk 2]
```diff
@@ -258,8 +277,20 @@ while [ -n "${1}" ]; do
     "--disable-go") NETDATA_DISABLE_GO=1 ;;
     "--enable-ebpf") NETDATA_ENABLE_EBPF=1 ;;
     "--disable-cloud")
-      NETDATA_DISABLE_CLOUD=1
-      NETDATA_CONFIGURE_OPTIONS="${NETDATA_CONFIGURE_OPTIONS//--disable-aclk/} --disable-aclk"
+      if [ -n "${NETDATA_REQUIRE_CLOUD}" ] ; then
+        echo "Cloud explicitly enabled, not re-enabling it"
```
Comment (prologic) [inline netdata-installer.sh]:
```suggestion
        echo "Cloud explicitly enabled, ignoring --disable-cloud option"
```

Confusing :)

File: netdata-installer.sh [hunk 3]
```diff
@@ -258,8 +277,20 @@ while [ -n "${1}" ]; do
...
+      else
+        NETDATA_DISABLE_CLOUD=1
+        NETDATA_CONFIGURE_OPTIONS="${NETDATA_CONFIGURE_OPTIONS//--disable-aclk/} --disable-aclk"
+      fi
+      ;;
+    "--require-cloud")
+      if [ -n "${NETDATA_DISABLE_CLOUD}" ] ; then
+        echo "Cloud explicitly disabled, not re-enabling it"
```
Comment (prologic) [inline netdata-installer.sh]:
```suggestion
        echo "Cloud explicitly disabled, ignoring --require-cloud option"
```

Confusing :)

File: netdata-installer.sh [hunk 4]
```diff
@@ -480,9 +511,19 @@ bundle_libmosquitto() {
       run_ok "libmosquitto built and prepared."
     else
       run_failed "Failed to build libmosquitto. The install process will continue, but you will not be able to connect this node to Netdata Cloud."
```
Comment (prologic) [inline netdata-installer.sh]: Please update messaging :)

File: netdata-installer.sh [hunk 5]
```diff
@@ -480,9 +511,19 @@ bundle_libmosquitto() {
...
+      if [ -n "${NETDATA_REQUIRE_CLOUD}" ] ; then
+        exit 1
```
Comment (prologic) [inline netdata-installer.sh]: See above

File: netdata-installer.sh [hunk 6]
```diff
@@ -480,9 +511,19 @@ bundle_libmosquitto() {
...
+      else
+        defer_error "Failed to build libmosquitto. The install process will continue, but you will not be able to connect this node to Netdata Cloud."
+      fi
     fi
   else
     run_failed "Unable to fetch sources for libmosquitto. The install process will continue, but you will not be able to connect this node to Netdata Cloud."
```
Comment (prologic) [inline netdata-installer.sh]: Same here

**Comment by prologic** [comment]
Hmmm I can't build from this PR (_at all_):

```sh
prologic@Jamess-iMac
Wed Mar 11 08:40:10
~/NetData/netdata
 (aclk-cleanup) 0
$ dki -t --rm -v "$PWD":/netdata -w /netdata alpine /bin/sh -c 'apk --no-cache -U add bash && ./packaging/installer/install-required-packages.sh --dont-wait --non-interactive netdata-all && ./netdata-installer.sh --dont-wait --dont-start-it --require-cloud || /bin/sh'
.
.
.
In file included from aclk/aclk_lws_wss_client.c:1:
aclk/aclk_lws_wss_client.h:4:10: fatal error: libwebsockets.h: No such file or directory
    4 | #include <libwebsockets.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:4334: aclk/aclk_lws_wss_client.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from aclk/mqtt.c:6:
aclk/aclk_lws_wss_client.h:4:10: fatal error: libwebsockets.h: No such file or directory
    4 | #include <libwebsockets.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:4334: aclk/mqtt.o] Error 1
make[2]: Leaving directory '/netdata'
make[1]: *** [Makefile:4932: all-recursive] Error 1
make[1]: Leaving directory '/netdata'
make: *** [Makefile:2690: all] Error 2
 FAILED

/netdata #
```

**Comment by prologic** [comment]
Why isn't this CI failing:

> Builder / Build (alpine:3.11) (pull_request) Successful in 2m

?

**Comment by prologic** [comment]
There we go:

> Builder / Build (alpine:3.9) (pull_request) Failing after 1m — Build (alpine:3.9)

CI _should_ fail when there is a failure :)

File: netdata-installer.sh (hunk 1 above)
Comment (Ferroin) [reply netdata-installer.sh]: I'm not 100% certain that this achieves what my intent was. I'm trying to deal with the fact that `echo` did not, for some reason, expand `\n` properly in the variable, which in and of itself didn't have it expanded properly for reasons I don't really understand.

Comment (prologic) [reply netdata-installer.sh]: echo by default will not expand or interpret blackslash sequences Unless you use the `-e` option:

> -e     enable interpretation of backslash escapes

**Comment by Ferroin** [comment]
Due to the issues with the handling of the feature flag, I'm closing this PR and splitting it into two separate PR's, one for the deferred error reporting (which should be immediately mergable), and one for the new installer flag (which will be blocked by #8378).

File: netdata-installer.sh (hunk 1 above)
Comment (Ferroin) [reply netdata-installer.sh]: That's only some versions of `echo`, not all support it (FreeBSD's implementation in particular does not).

**Pull request closed without merging**